### ISSUE TITLE
Don't pay vote Debug logging cost in normal operation

### ISSUE THREAD

**Issue by ValarDragon** [cometbft/cometbft#2847]
### Summary

Currently we spend lots of RAM allocation in Debug logs across votes. Here are some profiles of a query serving full node in its operation over 1-2 hours. We see:

addVote's StringShort calls taking 5GB of allocation (compared to the state machines total 48GB):
![Screenshot from 2024-04-18 20-55-22](https://github.com/cometbft/cometbft/assets/6440154/b4cfe07b-aac6-4df1-9f2d-c2255e334b62)
These are both debug line calls:
- https://github.com/cometbft/cometbft/blob/d27732ad9a992e321e440417a2a8220f90fe0044/internal/consensus/state.go#L2289
- https://github.com/cometbft/cometbft/blob/d27732ad9a992e321e440417a2a8220f90fe0044/internal/consensus/state.go#L2378

Peerstate.setHasVote taking 15.4GB, this is entirely in debug logging:
https://github.com/cometbft/cometbft/blob/d27732ad9a992e321e440417a2a8220f90fe0044/internal/consensus/reactor.go#L1444-L1449

### Problem Definition

This is certainly adding notable load to the system, and its not even providing utility most of the time!

### Proposal

I'm not sure what the most well abstracted way to do the logic for "if debug mode, run this log logic" is. Just wanted to point out that there is a lot were paying right now

**Comment by cason** [comment]
I think this can do the job: https://github.com/cometbft/cometbft/blob/d27732ad9a992e321e440417a2a8220f90fe0044/libs/log/lazy.go#L14-L19

**Comment by ValarDragon** [comment]
Heres a screenshot for SetHasVote, sorry for not including that in OP: 
![image](https://github.com/cometbft/cometbft/assets/6440154/ce4046fb-20fb-485d-a404-b0bbbb14081e)

**Comment by cason** [comment]
But for the first 2 examples, we can do what we already have for block hashes, with `LazyBlockHash`. 

In both cases it is a vote set. The logic is that we pass the vote set (reference), and only call the `StringShort()` if the logger produces the log message.

**Comment by andynog** [comment]
I agree that there are too many allocations in this logic, especially on `consensus/state`. Unfortunately the current logging logic allocates heap even if debug mode is not enabled. 

I've played around and implemented some logic on a branch to replace that, adding a new method on the `Logger` called `DebugOn` that returns a `bool` if the debug mode is enabled. We have many loggers implemented on this interface and I implemented the logic only on `filter` logger.

And then I replaced all the `logger.Debug` logic in the `consensus/state` to check first `if logger.DebugOn()` then you log. It works, the `logger.Debug()` is not invoked and hopefully these allocations won't happen. But I didn't have time to profile it. 

What do you think about this approach @ValarDragon and @cason ?

My branch can be found here and the commit is -> https://github.com/cometbft/cometbft/commit/d6dfcf1bb439f1f944d21c39c2e33e07cdf59004

**Comment by ValarDragon** [comment]
This sounds great to me, I really like it! Can you add it to these two spots:
- https://github.com/cometbft/cometbft/blob/d27732ad9a992e321e440417a2a8220f90fe0044/internal/consensus/reactor.go#L1444-L1449
- https://github.com/cometbft/cometbft/blob/main/p2p/conn/connection.go#L361-L362

Then can help with getting a mainnet profile for it! With the cumulative changes here. I think this will be a big performance difference!

**Comment by andynog** [comment]
Cool, I'll look into these other places too. If you can do a profile that would be amazing to validate the performance improvement

**Comment by andynog** [comment]
Hi @ValarDragon I've added the additional places you mentioned above, I've also added a few more that might help.

Not sure how you are going to profile this but I pushed a draft PR that it's easier to visualize the changes #2989 

Please let me know if there's anything else you need in order to be able to test this. Thanks again !

**Comment by ValarDragon** [comment]
Commented out the spot instances in Osmosis fork for noticeable performance gains / reductions in gossip routines and consensus mutex!

**Comment by melekes** [comment]
Here is an idea: during our release process, we produce a separate binary without all debug calls (`find /path/to/folder -type f -exec sed -i '/^log\.Debug/d' {} +`).

**Comment by cason** [comment]
This is a good idea, but only works for single-line debug commands. In any case, we use `logger`.

**Comment by sergio-mena** [comment]
Where is our good old `#define` ? 😞

**Comment by cason** [comment]
There is a super expensive (for who is doing it) but simple idea: if logger is debug, then code for the debug calls.

**Comment by andynog** [comment]
Maybe we can encapsulate the debug log and use build tags to disable it? Something along the lines

```
// +build !debug

package main

import (
    log "github.com/sirupsen/logrus"
)

func debugLog(logger *log.Logger, msg string) {
    // Do nothing, debug logging is disabled
}
```

**Comment by andynog** [comment]
got some ideas on how this might work. When I have a bit of time, will try to implement that

**Comment by cason** [comment]
Not sure if we should close this issue, although @andynog's solution is a way to go. The option to move to another provider (e.g., `slog`) is still in the table.

**Comment by andynog** [comment]
Hi @ValarDragon, should we leave this open ? We've implemented a solution in #4365 that in certain ways is a solution to the problem.

Please let us know. Thanks !
